activerecord/lib/active_record/relation/predicate_builder.rb

```diff
@@ -79,7 +81,20 @@ def expand_from_hash(attributes)
           if value.is_a?(Hash) && !table.has_column?(key)
             associated_predicate_builder(key).expand_from_hash(value)
           else
-            build(table.arel_attribute(key), value)
+            predicates = build(table.arel_attribute(key), value)
+
+            # Grouping NOT conditions that are of polymorphic associations
+            # in order to provide their proper inversion.
+            if of_inverted_clause &&
+              ((value.is_a?(AssociationQueryValue) &&
+                value.associated_table.polymorphic_association?) ||
+                  (value.is_a?(PolymorphicArrayValue) &&
+                    !predicates.is_a?(Arel::Nodes::Grouping)))
```

Comment:
The goal was to affect the least possible number of classes and methods. This is due to the fact that there is already the proper way (https://github.com/rails/rails/pull/26058) to solve the problem, but it can not be immediately implemented. I mean, sometime the problem will be solved by changing `where.not()` behaviour, so what is done in this code will not be necessary and will be removed. But it will be necessary until the problem is solved and until there is the need to maintain both (old and new) versions of `where.not()` (if you don't understand what I am talking about, see the above link).  So, I see this PR as a temporary solution and I don't see another way how to make it better without affecting less code.